x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java

```diff
@@ -151,13 +165,110 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
             concreteIndices.put(ir.getIndexName(), ir.getIndexMode());
+
+            String clusterAlias = parseClusterAlias(indexExpression);
+            clusterAndResolvedIndices.compute(clusterAlias, (k, v) -> {
+                if (v == null) {
+                    return new StringBuilder().append(indexExpression);
+                } else {
+                    return v.append(',').append(indexExpression);
+                }
+            });
+        }
+        if (executionInfo != null) {
+            updateExecutionInfoWithFieldCapsResults(executionInfo, clusterAndResolvedIndices, fieldCapsResponse.getFailures());
         }
         return IndexResolution.valid(new EsIndex(indexPattern, rootFields, concreteIndices));
     }
 
+    // visible for testing
+    static void updateExecutionInfoWithFieldCapsResults(
+        EsqlExecutionInfo executionInfo,
+        Map<String, StringBuilder> clusterAndResolvedIndices,
+        List<FieldCapabilitiesFailure> failures
+    ) {
+        Set<String> clustersWithoutFieldCapsResponses = new HashSet<>(executionInfo.clusterAliases());
+        for (Map.Entry<String, StringBuilder> entry : clusterAndResolvedIndices.entrySet()) {
+            final String clusterAlias = entry.getKey();
+            final EsqlExecutionInfo.Cluster cluster = executionInfo.getCluster(clusterAlias);
+            assert cluster != null : "All cluster aliases should have already been inserted into ExecutionInfo. Missing: " + clusterAlias;
+            executionInfo.swapCluster(
+                clusterAlias,
+                // overwrite the existing unresolved index expression with the resolved indices
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(cluster).setIndexExpression(entry.getValue().toString()).build()
+            );
+            clustersWithoutFieldCapsResponses.remove(clusterAlias);
+        }
+
+        /*
+         * These are clusters in the original request that are not present in the field-caps response. They were
+         * specified with an index or indices that do not exist, so the search on that cluster is done.
+         * Mark it as complete with 0 shards searched and 0 took time.
+         */
+        for (String c : clustersWithoutFieldCapsResponses) {
+            executionInfo.swapCluster(
+                c,
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
+                    .setTook(new TimeValue(0))
+                    .setTotalShards(0)
+                    .setSuccessfulShards(0)
+                    .setSkippedShards(0)
+                    .setFailedShards(0)
+                    .build()
+            );
+        }
+
+        if (failures != null) {
+            Set<String> unavailableClusters = new HashSet<>();
+            for (FieldCapabilitiesFailure failure : failures) {
+                if (ExceptionsHelper.isRemoteUnavailableException(failure.getException())) {
```

Comment:
If we (ES|QL) fully fork FieldCaps that would very doable. I'm leery of changing the core field-caps since that is used in other contexts and this PR is large enough that I wouldn't want to take it on in this ticket. I'll put a code comment here about that and create a GH issue to add that feature to either field-caps or EsqlResolveFieldsAction.

Reply:
Issue created: https://github.com/elastic/elasticsearch/issues/113394